File: Python/examples/american-option.py
# ---
# jupyter:
#   jupytext:
#     formats: py:percent
#     text_representation:
#       extension: .py
#       format_name: percent
#       format_version: '1.3'
#       jupytext_version: 1.4.2
#   kernelspec:
#     display_name: Python 3
#     language: python
#     name: python3
# ---

# %% [markdown]
# # American options
#
# Copyright (&copy;) 2004, 2005, 2006, 2007 StatPro Italia srl
#
# This file is part of QuantLib, a free-software/open-source library
# for financial quantitative analysts and developers - https://www.quantlib.org/
#
# QuantLib is free software: you can redistribute it and/or modify it under the
# terms of the QuantLib license.  You should have received a copy of the
# license along with this program; if not, please email
# <quantlib-dev@lists.sf.net>. The license is also available online at
# <https://www.quantlib.org/license.shtml>.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the license for more details.

# %%
import QuantLib as ql
import pandas as pd

# %% [markdown]
# ### Global parameters

# %%
todaysDate = ql.Date(15, ql.May, 1998)
ql.Settings.instance().evaluationDate = todaysDate

# %%
interactive = "get_ipython" in globals()

# %% [markdown]
# ### Option construction

# %%
exercise = ql.AmericanExercise(todaysDate, ql.Date(17, ql.May, 1999))
payoff = ql.PlainVanillaPayoff(ql.Option.Put, 40.0)

# %%
option = ql.VanillaOption(payoff, exercise)

# %% [markdown]
# ### Market data

# %%
underlying = ql.SimpleQuote(36.0)
dividendYield = ql.FlatForward(todaysDate, 0.00, ql.Actual365Fixed())
volatility = ql.BlackConstantVol(todaysDate, ql.TARGET(), 0.20, ql.Actual365Fixed())
riskFreeRate = ql.FlatForward(todaysDate, 0.06, ql.Actual365Fixed())

# %%
process = ql.BlackScholesMertonProcess(
    ql.QuoteHandle(underlying),
    ql.YieldTermStructureHandle(dividendYield),
    ql.YieldTermStructureHandle(riskFreeRate),
    ql.BlackVolTermStructureHandle(volatility),
)

# %% [markdown]
# ### Pricing
#
# We'll collect tuples of method name, option value, and estimated error from the analytic formula.

# %%
results = []

# %% [markdown]
# #### Analytic approximations

# %%
option.setPricingEngine(ql.BaroneAdesiWhaleyApproximationEngine(process))
results.append(("Barone-Adesi-Whaley", option.NPV()))

# %%
option.setPricingEngine(ql.BjerksundStenslandApproximationEngine(process))
results.append(("Bjerksund-Stensland", option.NPV()))

# %% [markdown]
# #### Finite-difference method

# %%
timeSteps = 801
gridPoints = 800

# %%
option.setPricingEngine(ql.FdBlackScholesVanillaEngine(process, timeSteps, gridPoints))
results.append(("finite differences", option.NPV()))


# %% [markdown]
# #### Li, M. QD+ American engine

# %%
option.setPricingEngine(ql.QdPlusAmericanEngine(process))
results.append(("QD+", option.NPV()))


# %% [markdown]
# #### Leif Andersen, Mark Lake and Dimitri Offengenden high performance American engine

# %%
option.setPricingEngine(
    ql.QdFpAmericanEngine(process, ql.QdFpAmericanEngine.accurateScheme())
)
results.append(("QD+ fixed point", option.NPV()))


# %% [markdown]
# #### Binomial method

# %%
timeSteps = 801

# %%
for tree in ["JR", "CRR", "EQP", "Trigeorgis", "Tian", "LR", "Joshi4"]:
    option.setPricingEngine(ql.BinomialVanillaEngine(process, tree, timeSteps))
    results.append(("Binomial (%s)" % tree, option.NPV()))

# %% [markdown]
# ### Results

# %%
df = pd.DataFrame(results, columns=["Method", "Option value"])
df.style.hide(axis="index")

# %% [markdown]
# The following displays the results when this is run as a Python script (in which case the cell above is not displayed).

# %%
if not interactive:
    print(df)

File: R/demo/american-option.R

suppressMessages(library(QuantLib))

## global data
calendar <- TARGET()

settlementDate <- Date(15, "May", 1998)
settlementDate <- Calendar_adjust(calendar, settlementDate)

fixingDays <- 3
settlementDays <- 3
todaysDate <- Calendar_advance(calendar, settlementDate, -fixingDays, "Days")
invisible(Settings_instance()$setEvaluationDate(d=todaysDate))

cat('Today          : ', todaysDate$`__str__`(), "\n")
cat('Settlement Date: ', settlementDate$`__str__`(), "\n")


## option
exercise = AmericanExercise(todaysDate, Date(17, "May", 1999))
payoff = PlainVanillaPayoff("Put", 40.0)
option = VanillaOption(payoff, exercise)


# ### Market data

# %%
underlying = SimpleQuote(36.0)
dividendYield = FlatForward(todaysDate, 0.00, Actual365Fixed())
volatility = BlackConstantVol(todaysDate, calendar, 0.20, Actual365Fixed())
riskFreeRate = FlatForward(todaysDate, 0.06, Actual365Fixed())

# %%
process = BlackScholesMertonProcess(
    QuoteHandle(underlying),
    YieldTermStructureHandle(dividendYield),
    YieldTermStructureHandle(riskFreeRate),
    BlackVolTermStructureHandle(volatility)
)

# ### Pricing
# We'll collect tuples of method name, option value, and estimated error from the analytic formula.

# %%
results = NULL

# #### Analytic approximations

option$setPricingEngine(BaroneAdesiWhaleyApproximationEngine(process))
results = rbind(results, data.frame("method"="Barone-Adesi-Whaley", "NPV"=option$NPV()))

option$setPricingEngine(BjerksundStenslandApproximationEngine(process))
results = rbind(results, data.frame("method"="Bjerksund-Stensland", "NPV"=option$NPV()))


# #### Finite-difference method

timeSteps = 801
gridPoints = 800

option$setPricingEngine(FdBlackScholesVanillaEngine(process, timeSteps, gridPoints))
results = rbind(results, data.frame("method"="finite differences", "NPV"=option$NPV()))


# #### Binomial method

timeSteps = 801

# %%
listEngineFuncsNames = lsf.str("package:QuantLib")
listEngineFuncsNames = listEngineFuncsNames[grepl(pattern = "^Binomial", x = listEngineFuncsNames)]
listEngineFuncsNames = listEngineFuncsNames[grepl(pattern = "VanillaEngine$", x = listEngineFuncsNames)]

for (engineFuncName in listEngineFuncsNames) {
  eval(parse(text=paste0("option$setPricingEngine(", engineFuncName, "(process, timeSteps))")))
  results = rbind(results, data.frame("method"=engineFuncName, "NPV"=option$NPV()))
}


# ### Results
print(results)
